### ISSUE TITLE
z-order not working correctly when docking new UI windows

### ISSUE THREAD

**Issue by Colpocorto** [openMSX/openMSX#1624]
It seems that the z-order of a docking target isn't taken into account.

This causes windows to accidentally "disappear" if dropped onto a hidden window.
The following video shows how the docking destinations are active for a OpenMSX window that is behind a Google Chrome window.

https://youtu.be/gkiJEfkHHJ0

Version of OpenMSX: 19.1-909
OS: Windows 10 64bit Intel

**Comment by MBilderbeek** [comment]
I'm not sure whether the Dear ImGUI framework can know that the window of another process is in between 2 windows of Dear ImGUI....

**Comment by MBilderbeek** [comment]
@ocornut can you comment? Is there any hope we could do something about this?

**Comment by ocornut** [comment]
This requires the backend to fully support the ImGuiBackendFlags_HasMouseHoveredViewport flag. The SDL2 backend doesn’t because it lacks the correct information from SDL2z

We could, should perhaps, implement the same win32-only workaround we use for the glfw backend, hooking WndProc to handle input-transparent windows. I can try looking into it, but note that the fix would only work for Windows.

If you use imgui_impl_win32.cpp when running on Windows it would also solve your problem, but may be more work.

**Comment by ej-sanmartin** [comment]
How open would openMSX be to migrating from SDL2 to SDL3? I have been contributing and looking deeply to SDL and see they do have an internal implementation (in Cocoa backend, so not widely implemented) of getting which window is on top, if this is the information needed @ocornut to make this backend fully support `ImGuiBackendFlags_HasMouseHoveredViewport` flag.

I'd be willing to open the issue with SDL, implement and get more info on the state of an API like `SDL_Window* SDL_GetWindowFromGlobalPoint(float x, float y)` on SDL's end. The only limitation I could think of is that it would only be able to get SDL windows and not a Chrome window but I think that should be enough info needed for input routing.

I understand this would be significant effort on openMSX's part and there may be other pressing features on the roadmap, but given this issue is affecting this bug and potentially #2058 it would be something to consider.

---

Edit: Just want to make explicit, if this is implemented in SDL we'd have to also update imgui, where relevant, that this info can be reliably obtained now

**Comment by MBilderbeek** [comment]
We are very open to migrate to SDL3. We are even planning to. 
What's holding us back right now is that it isn't in the GitHub Actions Ubuntu image and the work for the windows visual studio build.

Help most welcome.

**Comment by ej-sanmartin** [comment]
I'd be happy to help with migration efforts to SDL3 once that's started!

On the blockers:

- Looked into this and GitHub Actions’ `ubuntu-latest` is `Ubuntu 24.04`, which doesn’t reliably have `libsdl3-dev` in its default repos, but newer Ubuntu releases like 25.04 (Plucky) and 26.04 (Resolute) do include a `libsdl3-dev` package in their universe archives.  We could open an issue at https://github.com/actions/runner-images to support a newer Ubuntu image that includes `libsdl3-dev`, but they’d likely decline for now since they [somewhat recently rolled out the `Ubuntu-latest` pointing to the 24.04 image](https://github.com/actions/runner-images/issues/10636). I imagine updating to newer Ubuntu releases is something they'd do later naturally though idk their timeline on that. Thus, the practical solution today is using SDL’s official GitHub Action [libsdl-org/setup-sdl](https://github.com/libsdl-org/setup-sdl), which builds/installs SDL3 in CI independently of the runner image.
- To clarify, windows visual studio build issues are [these](https://github.com/openMSX/openMSX/issues?q=is%3Aissue%20state%3Aopen%20windows%20visual%20studio)?

**Comment by MBilderbeek** [comment]
So, you already solved point 1! Great.
Point 2: as you can see in our GitHub Actions, we also build with Visual Studio. Unfortunately, it always is a bit of a trouble to migrate the build files of a new library (and sometimes even also an upgrade of an existing library) into the openMSX build files which are VS specific. 

If you can simply make a PR with both these things working and the resulting binaries work fine, then you have just accomplished the migration to SDL3!

Let me know if you need more help and explanations to move forward.

**Comment by m9710797** [comment]
@MBilderbeek I wouldn't consider that a "SDL3 migration", but it's a good first step.
SDL3 offers several new features that I'd like to use. E.g. builtin support for mixing and resampling multiple audio streams (would allows us to git rid of our resampling). Or native file dialogs (would allow us to remove src/3rdparty/ImGuiFileDialog, which people sometimes complain about), ...

BUT isn't it yet too soon to already _require_ SDL3? Ubuntu 25.04 was release less than one year ago. E.g. we currently still allow to build with gcc-12 (released May 2022). Though soon-ish I'd like to bump that requirement to gcc-13 (released April 2023). **_IF_** we're fine with SDL3 as a required dependency, then I'd also like to immediately require gcc-14 (released May 2024).

About the library building situation:
@mthuurne Not trying to push, just asking. Did I understand correctly that you planned to work on using a package manager (like conan) to make this easier? What's the current status?

**Comment by MBilderbeek** [comment]
The first step is to move to SDL3 and get everything back to work as it is now. 
And then we can make (better) use of it.

That first step is usually the most difficult, considering the things I mentioned.

Anyways, SDL3 was released/recommended February 2025, that's almost a year ago... I'm not sure what's wise here, but we've always been a bit 'at the front' of things, right? :)

**Comment by ej-sanmartin** [comment]
@m9710797 I agree, the migration would be a bit more involved (potential method signature updates, ensuring existing SDL2 workarounds are removed and we get behavior parity with SDL3, for example).

I'm not next to my main machine, but I estimate this touches a large majority of the codebase. Luckily 95% of migration is mostly mechanical (guide has some advice https://github.com/libsdl-org/SDL/blob/main/docs/README-migration.md).


I don’t think it’s too early from a technical standpoint, SDL3 has had a stable release Jan 2025 with v3.2.0, and we’re already hitting real SDL2 limitations and bugs that can’t be cleanly fixed there. Migrating sooner keeps the diff smaller and avoids accumulating more SDL2-specific workarounds.

I agree with the approach of first getting a 1:1 working SDL3 build before taking advantage of new APIs, and I’d suggest opening a dedicated issue to track the SDL2 -> SDL3 migration so the scope and blockers are explicit (easier tracking for myself, seems a lot of context in this may be across multiple issues). To ease the transition, we could also consider using sdl12-compat, a compatibility layer, allowing us to support both environments during the migration without blocking progress.

Last note, requiring gcc-14 seems fair considering the justifications here for SDL3 though I wouldn't consider it a blocker to this migration. To minimize scope, we can ensure stability there after this migration or treat it as independent to this issue.
